storage/isolation_manager.rs

```diff
@@ -241,297 +358,317 @@ struct Timeline {
+            window.increment_readers();
+        };
     }
 
     fn remove_reader(&self, reader_sequence_number: SequenceNumber) {
-        let read_window = self.get_window(reader_sequence_number);
-        self.remove_window_reader(reader_sequence_number, &read_window);
+        if let Some(window) = self.find_window_for_reader(reader_sequence_number) {
+            debug_assert!(window.get_readers() >= 0);
+            let _readers_remaining = window.decrement_readers();
+            if _readers_remaining == 0 {
+                self.may_free_windows();
+            }
+        };
     }
 
-    fn remove_window_reader(
+    fn find_window_for_reader(
         &self,
-        reader_sequence_number: SequenceNumber,
-        reader_window: &TimelineWindow<TIMELINE_WINDOW_SIZE>,
-    ) {
-        debug_assert!(reader_window.contains(reader_sequence_number));
-        let _readers_remaining = reader_window.decrement_readers();
-
-        // TODO: clean up windows that no longer need to be retained in memory.
-        //      --> see task
-    }
-
-    fn last_window(
-        windows: &VecDeque<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>>,
-    ) -> Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>> {
-        debug_assert!(!windows.is_empty());
-        windows[windows.len() - 1].clone()
+        sequence_number: SequenceNumber,
+    ) -> Option<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>> {
+        let (_, windows) = &*self.next_window_and_windows.read().unwrap_or_log();
+        windows
+            .iter()
+            .rev()
+            .find(|window| {
+                let start_seq = window.starting_sequence_number();
+                start_seq.is_some() && start_seq.unwrap().number() <= sequence_number.number()
+            })
+            .cloned()
     }
 
-    fn get_or_create_window(&self, sequence_number: SequenceNumber) -> Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>> {
-        self.try_get_window(sequence_number).clone().unwrap_or_else(|| self.create_windows_to(sequence_number))
+    fn collect_concurrent_windows(
+        &self,
+        start_sequence_number: SequenceNumber,
+        end_relative_index: i64,
+    ) -> (Vec<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>>, i64) {
+        let (next_window, windows) = &*self.next_window_and_windows.read().unwrap_or_log();
+        let relative_index_of_window_0 = *next_window - (windows.len() * TIMELINE_WINDOW_SIZE) as i64;
+        assert!(end_relative_index >= relative_index_of_window_0);
+        let end_lies_in_window = (end_relative_index - relative_index_of_window_0) as usize / TIMELINE_WINDOW_SIZE;
+        let mut concurrent_windows: Vec<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>> = Vec::new();
+
+        let window_after = (0..windows.len()).find(|i| {
+            let window_start_seq = windows.get(*i).unwrap().starting_sequence_number().unwrap().number();
+            window_start_seq > start_sequence_number.number()
+        });
```

Comment:
This scan would go away and just re-use the `try_get_window(relative index  +  1)` if we had a relative index right? What if we introduce a `IndexedSequenceNumber`, which is a pair of relative index and sequence number and always returned from the durability service on durable write (unless you don't provide a sequenced record/ do an unsequenced write)?